internal/auth/internal/types/traffic_permissions.go

```diff
@@ -31,114 +31,224 @@ func RegisterTrafficPermissions(r resource.Registry) {
 		Proto:    &pbauth.TrafficPermissions{},
 		Scope:    resource.ScopeNamespace,
 		Validate: ValidateTrafficPermissions,
+		Mutate:   MutateTrafficPermissions,
 	})
 }
 
+func MutateTrafficPermissions(res *pbresource.Resource) error {
+	var tp pbauth.TrafficPermissions
+
+	if err := res.Data.UnmarshalTo(&tp); err != nil {
+		return resource.NewErrDataParse(&tp, err)
+	}
+
+	changed := false
+
+	parentTenancy := &pbresource.Tenancy{}
+	if t := res.Id.Tenancy; t != nil {
+		parentTenancy = t
+	}
+
+	for _, p := range tp.Permissions {
+		for _, s := range p.Sources {
+			if updated := normalizedTenancyForSource(s, parentTenancy); updated {
+				changed = true
+			}
+		}
+	}
+
+	if !changed {
+		return nil
+	}
+
+	return res.Data.MarshalFrom(&tp)
+}
+
+func normalizedTenancyForSource(src *pbauth.Source, parentTenancy *pbresource.Tenancy) bool {
+	var changed bool
+
+	if t, c := defaultedSourceTenancy(src, parentTenancy); c {
+		src.Partition = t.Partition
+		src.Peer = t.PeerName
+		changed = true
+	}
+
+	for _, e := range src.Exclude {
+		if t, c := defaultedSourceTenancy(e, parentTenancy); c {
+			e.Partition = t.Partition
+			e.Peer = t.PeerName
+			changed = true
+		}
+	}
+
+	return changed
+}
+
+func defaultedSourceTenancy(s pbauth.SourceToSpiffe, parentTenancy *pbresource.Tenancy) (*pbresource.Tenancy, bool) {
+	if !isLocalPeer(s.GetPeer()) {
+		return nil, false
+	}
+
+	tenancy := pbauth.SourceToTenancy(s)
+
+	var changed bool
+	tenancy.PeerName, changed = firstNonEmptyString(tenancy.PeerName, parentTenancy.PeerName, resource.DefaultPeerName)
+
+	var c bool
+	tenancy.Partition, c = firstNonEmptyString(tenancy.Partition, parentTenancy.Partition, resource.DefaultPartitionName)
+	if c {
+		changed = true
+	}
+
+	return tenancy, changed
```

Comment:
```suggestion
	return tenancy, changedPeer || changedPartition || changedNamespace
```
maybe that reads a bit better?